Sara/mkt 755 semgrep secrets launch documentation (#1210)
returntocorp/semgrep-docs: docs/semgrep-secrets/getting-started.md

---
slug: /semgrep-secrets/getting-started
append_help_link: true
title: Getting started with Semgrep Secrets
hide_title: true
description: Set up secrets scanning in Semgrep to find and rotate valid leaked secrets.
tags:
  - Semgrep Secrets
toc_max_heading_level: 2
---

import MoreHelp from "/src/components/MoreHelp"

# Getting started with Semgrep Secrets

Detect and triage leaked secrets and credentials by scanning with Semgrep Secrets.

Save time by prioritizing which secrets to rotate based on whether they are valid (active and in-use) or not.

![Semgrep Secrets page](/img/secrets-page.png)
**_Figure._** Semgrep Secrets page. <br />

This document guides you through the following:

* Enabling Semgrep Secrets
* Receiving findings in GitHub PR comments
* Triaging findings

> To learn about how Semgrep Secrets detects secrets, see [<i class="fa-regular fa-file-lines"></i> Conceptual overview of Semgrep Secrets](/docs/semgrep-secrets/conceptual-overview).

Reviewer: If you agree to the title change comment on the other file, this would also change.

:::info Feature maturity
* This feature is in public beta.

> * You or your developers may encounter rough edges. For issues, reach out to [<i class="fa-regular fa-envelope"></i> support@semgrep.com](mailto:support@semgrep.com).

Reviewer: ```suggestion
* You or your developers may encounter bugs. For issues, reach out to [<i class="fa-regular fa-envelope"></i> support@semgrep.com](mailto:support@semgrep.com).
```

Suggestion here is to not use idioms that may not be as well understood

:::

## Supported developer environments and features

This section lists the source code managers (SCMs) that Semgrep Secrets supports.

| Source code manager        | Semgrep Secrets | PR or MR comments for valid secrets findings   |
| -------------------        | --------------- | ---------------------------------------------- |
| GitHub - Free and Team     | ✔️               | ✔️                                              |
| GitHub - Enterprise Cloud  | ✔️               | ✔️                                              |
| GitHub - Enterprise Server | ❌ (Coming soon)| ❌                                             |
| GitLab (any plan)          | ❌ (Coming soon)| ❌                                             |

## Enabling Semgrep Secrets

:::info Prerequisites
* Semgrep Secrets can only be enabled through Semgrep Cloud Platform (SCP). [<i class="fas fa-external-link fa-xs"></i> Create an account](https://semgrep.dev/login) to enable Semgrep Secrets.
* You have added or onboarded at least one repository to Semgrep Cloud Platform for scanning. See Starting a SAST and SCA scan on a remote repository.
:::

> 1. Contact the Semgrep Team by filling out the [<i class="fas fa-external-link fa-xs"></i> Semgrep Secrets Beta](https://get.semgrep.dev/secrets-beta-request.html) form.

Reviewer: ```suggestion
1. Contact the Semgrep team by filling out the [<i class="fas fa-external-link fa-xs"></i> Semgrep Secrets Beta](https://get.semgrep.dev/secrets-beta-request.html) form.
```

Author: Committing suggestion - we should probably formalize this somewhere (Prefer Semgrep team over Semgrep Team) - I'll make a Linear task for it.

Author: Committing suggestion - we should probably formalize this somewhere (Prefer Semgrep team over Semgrep Team) - I'll make a Linear task for it.

Author: Committing suggestion - we should probably formalize this somewhere (Prefer Semgrep team over Semgrep Team) - I'll make a Linear task for it.

### Severity

Severity is assigned based on how sensitive or crucial the exposed web service is. Possible values include:

* High
* Medium
* Low

### Validation

Refers to whether or not a secret is active and can be used to grant resources or authentication, or if a secret is inactive.

| Validation | Description |
| -----------  | ------------ |
| Confirmed valid | Semgrep made an API call using the secret and it returned an HTTP response of 200 or similar and granted authentication. |
| Confirmed invalid | Semgrep made an API call using the secret and it returned an HTTP response of 403 or similar. |
| Validation error | Semgrep made an API call but it returned and HTTP response of 400 or similar; a server error, such as a timeout, occurred. The Semgrep Team recommends manually reviewing the finding. |
| No validator | Semgrep does not perform any validation on this finding. You must manually review the finding. |

### Repository visibility

Refers to whether or not the repository is a public repository or private. This is detected through your source code manager.

:::info

> Semgrep supports visibility detection only for GItHub repositories of any plan.

Reviewer: ```suggestion
Semgrep supports visibility detection only for GitHub repositories of any plan.
```

:::info Feature maturity
* This feature is in public beta.
* You or your developers may encounter rough edges. For issues, reach out to [<i class="fa-regular fa-envelope"></i> support@semgrep.com](mailto:support@semgrep.com).
:::

:::tip Sign-up
To gain access to this feature, fill out the [<i class="fas fa-external-link fa-xs"></i> Semgrep Secrets Beta](https://get.semgrep.dev/secrets-beta-request.html) form.
:::

## Supported developer environments and features

This section lists the source code managers (SCMs) that Semgrep Secrets supports.

| Source code manager                                     | Semgrep Secrets | PR or MR comments for valid secrets findings |
| -------------------                                     | --------------- | ----------------------------------------------   |
| GitHub - Free and Team                                  | ✔️               | ✔️                                                |
| GitHub - Enterprise Cloud                               | ✔️               | ✔️                                                |

> | GitHub - Enterprise Server                              | ✔️               | ?                                                |

Reviewer 2: I think right now we only support Github.com

> | GitHub - Enterprise Server                              | ✔️               | ?                                                |

Author: Updated with "X (Coming soon)" for items we don't support yet.

| GitLab.com (hosted plans) - Free, Premium, and Ultimate | ✔️               | ✔️                                                |

> | GitLab - Self-Managed plans                             | ✔️               | ?                                                |

Author: Same question as the above for gitlab :)

| -------------------                                     | --------------- | ----------------------------------------------   |
| GitHub - Free and Team                                  | ✔️               | ✔️                                                |
| GitHub - Enterprise Cloud                               | ✔️               | ✔️                                                |
| GitHub - Enterprise Server                              | ✔️               | ?                                                |
| GitLab.com (hosted plans) - Free, Premium, and Ultimate | ✔️               | ✔️                                                |
| GitLab - Self-Managed plans                             | ✔️               | ?                                                |

## Enabling Semgrep Secrets

:::info Prerequisites
* Semgrep Secrets can only be enabled through Semgrep Cloud Platform (SCP). [<i class="fas fa-external-link fa-xs"></i> Create an account](https://semgrep.dev/login) to enable Semgrep Secrets.
* You have added or onboarded at least one repository to Semgrep Cloud Platform for scanning. See Starting a SAST and SCA scan on a remote repository.
:::

> 

Reviewer 2: There is a step before this which is to contact the Semgrep team to get Secrets activated (since it's not self serve yet). What would be the best way to indicate that?

Author: Added a first step. Resolving.

4. Optional: After enabling Semgrep Secrets, you can trigger a full scan manually through your CI provider or wait for your scheduled Semgrep full scan, typically daily.

## Scanning environments

A scan can be triggered in the following environments:

* You can start a scan **manually** from your CI provider if your CI provider supports running CI jobs on-demand. Refer to your CI provider's documentation for details.
* You can start a scan **from your CLI**. It is recommended to run CLI scans on non-mainline branches, such as feature branches.
* When you **open a PR or MR**, Semgrep automatically scans the branch.

### Scanning for secrets through a CI job

After following the steps in [Enabling Semgrep Secrets](#enabling-semgrep-secrets), all Semgrep scans in your CI jobs include secret scanning. There are no additional steps to take.

After each scan, your findings are displayed in **Semgrep Cloud Platform > Secrets**.

:::caution Troubleshooting
If you do not receive any secrets findings in your scans after enabling Semgrep Secrets, check the following:

* Check the job's log in your CI provider.

>     * Search for the following line or similar to ensure that a secrets scan was included as part of the scan: `Enabled products: Semgrep Code, Semgrep Supply Chain, Semgrep Secrets`

Reviewer 2: ```suggestion
    * Search for the following line or similar to ensure that a secrets scan was included as part of the scan: `Enabled products: Code, Supply Chain, Secrets`
```

Author: ![Snag_639f2c](https://github.com/returntocorp/semgrep-docs/assets/20766840/a1998330-4785-4931-8483-f6dc2f3e3b6d)
@raghavjain3 It's the exact CLI output I'm quoting here.

* An existing Semgrep Cloud Platform account.
* An existing Semgrep organization.
* Semgrep Secrets must be enabled in Semgrep Cloud Platform.
:::

You can scan for secrets from your local development machine by using the Semgrep CLI tool. 

1. Ensure you are logged in to Semgrep:
```
semgrep login
```
2. Run a scan:
```
semgrep ci
```
3. Once the scan has completed, you can view that repository's findings in Semgrep Cloud Platform by logging in then clicking **Projects**, then the number of Secrets findings under <span className="placeholder">REPOSITORY_NAME</span>. <br />
![The number of Secrets findings can be seen in the Projects page.](/img/secrets-projects.png#md-width)

Findings from local scans are differentiated from their remote counterparts through their slugs. Remote repositories are identified as <span className="placeholder">  ACCOUNT_NAME/REPOSITORY_NAME</span>, while local repositories are identified as <span className="placeholder">REPOSITORY_NAME</span>.

> ## Receiving findings in GitHub or GitLab through PR or MR comments

Reviewer 2: ```suggestion
## Receiving findings in GitHub through PR comments
```

**_Figure._** Semgrep Secrets finding in a PR comment.

### Findings in GitHub pull requests

Perform the following steps to receive Secrets findings as comments in GitHub PRs:

1. Follow the steps in [<i class="fa-regular fa-file-lines"></i> GitHub PR comments](/semgrep-cloud-platform/github-pr-comments/).
2. Inform [<i class="fa-regular fa-envelope"></i> support@semgrep.com](mailto:support@semgrep.com) that you want to enable this feature.

### Findings in GitLab merge requests

Perform the following steps to receive Secrets findings as comments in GitLab MRs:

1. Follow the steps in [<i class="fa-regular fa-file-lines"></i> GitLab MR comments](/semgrep-cloud-platform/gitlab-mr-comments/).
2. Inform [<i class="fa-regular fa-envelope"></i> support@semgrep.com](mailto:support@semgrep.com) that you want to enable this feature.

### Receiving other notifications and alerts

You can receive findings notifications through the following channels:

* [<i class="fa-regular fa-file-lines"></i> Slack](/semgrep-cloud-platform/slack-notifications)
* [<i class="fa-regular fa-file-lines"></i> Email](/semgrep-cloud-platform/email-notifications)

> * [<i class="fa-regular fa-file-lines"></i> Webhooks](/semgrep-cloud-platform/webhooks)

Reviewer 2: We can remove these for now
